syntax = "proto3";

package photo;

// Service for receiving photos from partner application
service PhotoService {
    // Upload a photo with all metadata
    rpc UploadPhoto(PhotoRequest) returns (PhotoResponse);
    
    // Stream multiple photos
    rpc UploadPhotoBatch(stream PhotoRequest) returns (BatchPhotoResponse);
    
    // Legacy endpoint for backwards compatibility
    rpc UploadPlacePhoto(PlacePhotoRequest) returns (PlacePhotoResponse);
    
    // Legacy batch endpoint
    rpc UploadPlacePhotoBatch(stream PlacePhotoRequest) returns (LegacyBatchPhotoResponse);
}

// GeoPoint message (matches Android Firestore GeoPoint)
message GeoPoint {
    double latitude = 1;
    double longitude = 2;
}

// New Photo request format (matches Android data class Photo)
message PhotoRequest {
    // Photo ID (optional, will be generated if empty)
    string photo_id = 1;
    
    // Media URIs (list of photo URLs/paths)
    repeated string media_uris = 2;
    
    // Thumbnail URL
    string thumbnail_url = 3;
    
    // Audio URL (optional)
    string audio_url = 4;
    
    // Description
    string description = 5;
    
    // AI-generated tags
    repeated string ai_tags = 6;
    
    // Category (e.g., "Nature", "Urban")
    string category = 7;
    
    // Location Data
    string location_name = 8;
    GeoPoint geo_point = 9;
    string geohash = 10;
    string continent = 11;
    
    // Author Data (Denormalized)
    string author_id = 12;
    string author_name = 13;
    string author_avatar = 14;
    
    // Visibility & Social
    string visibility = 15;  // "PUBLIC" or "GROUP"
    string group_id = 16;    // Optional, only if visibility is GROUP
    int32 like_count = 17;
    
    // Timestamp (milliseconds since epoch)
    int64 timestamp = 18;
    
    // Binary photo data (optional, for direct upload)
    bytes photo_data = 19;
    
    // Content type for photo_data (e.g., "image/jpeg")
    string content_type = 20;
    
    // Source app identifier
    string source_app = 21;
}

// New Photo response
message PhotoResponse {
    bool success = 1;
    string photo_id = 2;
    string error_message = 3;
    
    // URLs for uploaded media
    repeated string media_urls = 4;
    string thumbnail_url = 5;
    string audio_url = 6;
}

// Batch upload response (new format)
message BatchPhotoResponse {
    int32 success_count = 1;
    int32 failed_count = 2;
    repeated PhotoResponse responses = 3;
}

// ========== Legacy messages for backwards compatibility ==========

// Legacy Photo upload request
message PlacePhotoRequest {
    // Photo data (binary)
    bytes photo_data = 1;
    
    // Photo file name
    string filename = 2;
    
    // Content type (image/jpeg, image/png, etc.)
    string content_type = 3;
    
    // Place information
    PlaceInfo place_info = 4;
    
    // Source app identifier (e.g., "instagram_clone")
    string source_app = 5;
    
    // User ID from source app (optional)
    string source_user_id = 6;
}

// Legacy Place information
message PlaceInfo {
    // Place name
    string name = 1;
    
    // Place address
    string address = 2;
    
    // City
    string city = 3;
    
    // Country (optional)
    string country = 4;
    
    // Coordinates
    double latitude = 5;
    double longitude = 6;
    
    // Google Place ID (if known)
    string google_place_id = 7;
    
    // Place types (e.g., "bar", "restaurant")
    repeated string types = 8;
}

// Legacy Upload response
message PlacePhotoResponse {
    // Success status
    bool success = 1;
    
    // Photo URL in MinIO
    string photo_url = 2;
    
    // Generated photo ID
    string photo_id = 3;
    
    // Error message (if any)
    string error_message = 4;
    
    // Matched Google Place ID (if found)
    string matched_place_id = 5;
}

// Legacy Batch upload response
message LegacyBatchPhotoResponse {
    // Number of successfully uploaded photos
    int32 success_count = 1;
    
    // Number of failed uploads
    int32 failed_count = 2;
    
    // Individual responses
    repeated PlacePhotoResponse responses = 3;
}
